Further refactor and simplify the ConGetSet API

j4james opened the pull request:
This is an attempt to simplify the `ConGetSet` interface down to the
smallest set of methods necessary to support the `AdaptDispatch` class.
The idea is that it should then be easier to implement that interface in
Windows Terminal, so we can replace the `TerminalDispatch` class with
`AdaptDispatch`.

This is a continuation of the refactoring started in #12247, and a
significant step towards #3849.

## Detailed Description of the Pull Request / Additional comments

The general idea was to give the `AdaptDispatch` class direct access to
the high-level structures on which it needs to operate. Some of these
structures are now passed in when the class is constructed (the
`Renderer`, `RenderSettings`, and `TerminalInput`), and some are exposed
as new methods in `ConGetSet` (`GetStateMachine`, `GetTextBuffer`, and
`GetViewport`).

Many of the existing `ConhostInternalGetSet` methods could easily then
be reimplemented in `AdaptDispatch`, since they were often simply
forwarding to methods in one of the above structures. Some were a little
more complicated, though, and require further explanation.

* `GetConsoleScreenBufferInfoEx`: What we were typically using this for
  was to obtain the viewport, although what we really wanted was the
  virtual viewport, which is now accessible via the `GetViewport`
  method. This was also used to obtain the cursor position and buffer
  width, which we can now get via the `GetTextBuffer` method.

* `SetConsoleScreenBufferInfoEx`: This was only really used for the
  `AdaptDispatch::SetColumns` implementation (for `DECCOLM` mode), and
  that could be replaced with `ResizeWindow`. This is a slight change in
  behaviour (it sizes the window rather than the buffer), but neither is
  technically correct for `DECCOLM`, so I think it's good enough for
  now, and at least it's consistent with the other VT sizing operations.

* `SetCursorPosition`: This could mostly be replaced with direct
  manipulation of the `Cursor` object (accessible via the text buffer),
  although again this is a slight change in behavior. The original code
  would also have made a call to `ConsoleImeResizeCompStrView` (which I
  don't think is applicable to VT movement), and would potentially have
  moved the viewport (not essential for now, but could later be
  supported by `DECHCCM`). It also called `VtIo::SetCursorPosition` to
  handle cursor inheritance, but that should only apply to
  `InteractDispatch`, so I've moved that to the
  `InteractDispatch::MoveCursor` method.

* `ScrollRegion`: This has been replaced by two simple helper methods in
  `AdaptDispatch` which better meet the VT requirements -
  `_ScrollRectVertically` and `_ScrollRectHorizontally`. Unlike the
  original `ScrollRegion` implementation, these don't generate
  `EVENT_CONSOLE_UPDATE_SCROLL` events (see #12656 for more details).

* `FillRegion`: This has been replaced by the `_FillRect` helper method
  in `AdaptDispatch`. It differs from the original `FillRegion` in that
  it takes a rect rather than a start position and length, which gives
  us more flexibility for future operations.

* `ReverseLineFeed`: This has been replaced with a somewhat refactored
  reimplementation in `AdaptDispatch`, mostly using the
  `_ScrollRectVertically` helper described above.

* `EraseAll`: This was previously handled by
  `SCREEN_INFORMATION::VtEraseAll`, but has now been entirely
  reimplemented in the `AdaptDispatch::_EraseAll` method.

* `DeleteLines`/`InsertLines`/`_modifyLines`: These have been replaced
  by the `_InsertDeleteLineHelper` method in `AdaptDispatch`, which
  mostly relies on the `_ScrollRectVertically` helper described above.

Finally there were a few methods that weren't actually needed in the
`ConGetSet` interface:

* `MoveToBottom`: This was really just a hack to get the virtual
  viewport from `GetConsoleScreenBufferInfoEx`. We may still want
  something like in the future (e.g. to support `DECVCCM` or #8879), but
  I don't think it's essential for now.

* `SuppressResizeRepaint`: This was only needed in `InteractDispatch`
  and `PtySignalInputThread`, and they could easily access the `VtIo`
  object to implement it themselves.

* `ClearBuffer`: This was only used in `PtySignalInputThread`, and that
  could easily access the buffer directly via the global console
  information.

* `WriteControlInput`: This was only used in `InteractDispatch`, and
  that could easily be replaced with a direct call to
  `HandleGenericKeyEvent`.

As part of these changes, I've also refactored some of the existing
`AdaptDispatch` code:

* `_InsertDeleteHelper` (renamed `_InsertDeleteCharacterHelper`) is now
  just a straightforward call to the new `_ScrollRectHorizontally`
  helper.

* `EraseInDisplay` and `EraseInLine` have been implemented as a series
  of `_FillRect` calls, so `_EraseSingleLineHelper` is no longer
  required.

* `_EraseScrollback` is a essentially a special form of scrolling
  operation, which mostly depends on the `TextBuffer::ScrollRows`
  method, and with the filling now provided by the new `_FillRect`
  helper.

* There are quite a few operations now in `AdaptDispatch` that are
  affected by the scrolling margins, so I've pulled out the common
  margin setup into a new `_GetVerticalMargins` helper method. This also
  fixes some edge cases where margins could end up out of range.

## Validation Steps Performed
There were a number of unit tests that needed to be updated to work
around functions that have now been removed, but these substitutions
were fairly straightforward for the most part.

The adapter tests were a different story, though. In that case we were
explicitly testing how operations were passed through to the `ConGetSet`
interface, but with more than half those methods now gone, a significant
rewrite was required.

I've tried to retain the crux of the original tests, but we now have to
validate the state changes on the underlying data structures, where
before that state would have been tracked in the `TestGetSet` mock. And
in some cases we were testing what happened when a method failed, but
since that scenario is no longer possible, I've simply removed those
tests.

I've also tried to manually test all the affected operations to confirm
that they're still working as expected, both in vttest as well as my own
test scripts.

Closes #12662

j4james commented:
There's more I'd still like to do to simplify `ConGetSet` even further, but I think this is good enough to move on to the `TerminalDispatch` replacement phase, which is more of a priority.

I know this is a big change, and there's a definite chance it might break something, but I do think it's worth the risk. Because once we've got a more complete VT implementation on the Terminal side, the pass-through mode starts to look like a realistic prospect.

lhecker commented:
FYI This week we're all a bit preoccupied with projects outside of regular Terminal work ("Fix Hack Learn").
Personally, I'll be on vacation next week, so I'll unfortunately only get a chance to take a look at this in about two weeks (unless it's already merged by then).

zadjii-msft commented:
Taking a closer look at this today. Some of the work that's happening in the #12515 and #12526 intersection will likely run across the congetset boundary, so IMO we should merge this in first, and let me & @miniksa deal with the fallout ourselves.

zadjii-msft commented on src/terminal/adapter/adaptDispatch.cpp:
```diff
@@ -394,60 +417,100 @@ bool AdaptDispatch::CursorRestoreState()
+            const auto data = OutputCell(*textBuffer.GetCellDataAt(sourcePos));
+            textBuffer.Write(OutputCellIterator({ &data, 1 }), targetPos);
+            source.WalkInBounds(sourcePos, walkDirection);
+        } while (target.WalkInBounds(targetPos, walkDirection));
```
Coreect me if I'm wrong here - if we had a buffer like:

```
1234567890
```

And the cursor was on the `1`, and we inserted 5 characters, we'd call this with `(textBuffer, { 0, 0, 80, 1 }, 5)`


<hr>

Actually nevermind me. That works perfectly fine. I thought we'd start reading from the section we were writing from, so it'd look like 

```
12341234
```

or something, but nah this works fine.

j4james replied:
I think I just stole that code from the existing `_CopyRectangle` function:
https://github.com/microsoft/terminal/blob/ce6ce50e7e9eeb2d7377527d9512da9d43674bc0/src/host/output.cpp#L107-L111

zadjii-msft commented on src/terminal/adapter/adaptDispatch.cpp:
```diff
@@ -1469,7 +1541,10 @@ bool AdaptDispatch::ForwardTab(const size_t numTabs)
     }
 
-    _pConApi->SetCursorPosition({ column, row });
+    cursor.SetXPosition(column);
+    cursor.SetDelay(false);
+    cursor.SetIsOn(true);
+    cursor.SetHasMoved(true);
```
Ah okay I found it. These come from: `SCREEN_INFORMATION::SetCursorPosition` 

https://github.com/microsoft/terminal/blob/0651d92dba09aaf13a807f1502a10bef159e4f5d/src/host/screenInfo.cpp#L1728-L1740

`IsOn` is always passed as true in `ApiRoutines::SetConsoleCursorPositionImpl`, so this right here is assuming that the console has focus.... that's probably okay. I don't entirely know the origin of that HAS_FOCUS check.

This may be served well by having a helper - I think I'm gonna have to introduce focus tracking even for conpty fairly soon.

j4james replied:
My impression was the the `HAS_FOCUS` check was just an optimization, and it shouldn't harm to set these flags anyway. The `TimerRoutine` that processes this stuff has it's own `HAS_FOCUS` check that prevents it triggering a notification if the window isn't focused.

https://github.com/microsoft/terminal/blob/ce6ce50e7e9eeb2d7377527d9512da9d43674bc0/src/host/CursorBlinker.cpp#L86-L89

If anything, I think the existing `HAS_FOCUS` check might cause it lose some move events, and could be a bug. But again I thought this stuff should probably be investigated further in a follow up issue. I didn't want to get too bogged down trying to resolve side issues like this if I could avoid it. As long as we aren't making thing worse in the meantime.

zadjii-msft reviewed (commented):
Wow, this is quite the read. I've got minor questions scattered about. I don't think I have any blocking concerns. Everything looks like it found a new home. 

Really looking forward to having one less layer to deal with after all these years.

zadjii-msft commented:
Another _project maintenance note:_ We've been largely focused on fixes that we can service back to 1.12, for syncing fixes with the OS build of the Terminal. This seems like it would firmly fit outside the realm of servicable fixes, so things that build off this may have a much harder time getting serviced to 1.12. 

Looking through the 1.14 list, there's like, two categories where this might matter:
* Alt buffer stuff - I don't think we're planning on bringing that back to 1.12 anymore, so ezpz
* the z-order / foreground / showhide stuff. This is... complicated
  * Plumbing `FocusIn`/`FocusOut` from the Terminal to conpty via InteractDispatch: before this PR, it'll require a new congetset method. After this PR, the body of that congetset method could certainly just get moved into InteractDispatch directly. So not the worst to have separate versions of that.
  * Same with the Show/Hide stuff, for `ConhostInternalGetSet::ShowWindow`. 


Okay, not that complicated. Definitely doable, annoying, but doable if we want the FG stuff in 1.12.

j4james commented on src/terminal/adapter/adaptDispatch.cpp:
```diff
@@ -1372,7 +1429,28 @@ bool AdaptDispatch::LineFeed(const DispatchTypes::LineFeedType lineFeedType)
+    else if (cursorPosition.Y > viewport.top)
+    {
+        // Otherwise we move the cursor up, but not past the top of the viewport.
+        const COORD newCursorPosition{ cursorPosition.X, cursorPosition.Y - 1 };
```
See the condition above (`cursorPosition.Y > viewport.top`). If `Y` is greater than the viewport top, then `Y-1` will still be greater than or equal to the top, which is still in range.

j4james commented on src/host/outputStream.cpp:
```diff
@@ -375,58 +214,6 @@ void ConhostInternalGetSet::WarningBell()
-    if (oldCursorPosition.Y > viewport.Top)
-    {
-        // Cursor is below the top line of the viewport
-        THROW_IF_NTSTATUS_FAILED(AdjustCursorPosition(screenInfo, newCursorPosition, TRUE, nullptr));
```
In this particular situation it wouldn't apply, because `ReverseLineFeed` shouldn't be panning the viewport when it moves the cursor. That said, When it reaches the top, it'll trigger a scroll of the buffer content, and that does do something similar to `InitializeCursorRowAttributes` for the revealed line. See the calls to `_FillRect` and `ResetLineRenditionRange` in the `_ScrollRectVertically` method.

j4james commented on src/terminal/adapter/adaptDispatch.cpp:
```diff
@@ -1126,7 +1147,49 @@ bool AdaptDispatch::SetCursorKeysMode(const bool applicationMode)
+        cursor.SetXPosition(0);
+        cursor.SetDelay(false);
+        cursor.SetIsOn(true);
+        cursor.SetHasMoved(true);
```
Yeah. I was considering that, but I thought maybe we could push that off for a follow up issue because I think this cursor handling needs further investigation.

If these calls are needed all the time, then why aren't they set automatically in the `SetPosition` method? If they aren't needed all the time, then we may need to be more selective about where they are used, and dumping these calls in a helper is just hiding the problem.

And note that Windows Terminal doesn't use these flags at all as far as I can tell, so I'm hoping it might be possible to do away with them altogether.

j4james replied:
FYI, I eventually decided you were right about this and added a helper for setting these flags. It does make things a bit tidier and gave me the opportunity to add a bit of documentation on what we are doing here (at least as I understand it). See commit 446002ed47d2e1a85f256362689435682caa5558.

j4james commented:
> * the z-order / foreground / showhide stuff. This is... complicated

I don't mind if you want to do those first. I only looked at them briefly, but I didn't think they'd be a huge problem to merge.

j4james commented:
I rebased to resolve some conflicts from PR #11264 (which were not a big deal), but that broke the adapter unit test build as a result of the dependencies that were removed in commit fd4fa664602d25fc2d3b52d82346e048b7328d66. I've added back a dependency on `RendererBase` to fix that, but I hope that's not going to be a problem for OneCore now (cc @DHowett).

lhecker commented on src/host/outputStream.cpp:
```diff
@@ -849,46 +351,21 @@ bool ConhostInternalGetSet::IsVtInputEnabled() const
-    auto* pRender = ServiceLocator::LocateGlobals().pRender;
-    if (pRender)
+    auto& screenInfo = _io.GetActiveOutputBuffer();
+    if (screenInfo.HasAccessibilityEventing() && screenInfo.IsActiveScreenBuffer())
```
The old `ConhostInternalGetSet::FillRegion` was notifying for the non-active screen buffer as well. I suppose this is a bug fix? (Because it does seem correct to not notify on the non-active one.)

lhecker commented on src/host/outputStream.cpp:
```diff
@@ -849,46 +351,21 @@ bool ConhostInternalGetSet::IsVtInputEnabled() const
-void ConhostInternalGetSet::UpdateSoftFont(const gsl::span<const uint16_t> bitPattern,
-                                           const SIZE cellSize,
-                                           const size_t centeringHint)
+void ConhostInternalGetSet::NotifyAccessibilityChange(const til::rect changedRect)
```
nit: If you can, pass `til::rect` by reference (and `til::point`/`til::size` by value).
Not really important for this PR though. I'll "standardize" the way we pass these types throughout the project in the future anyways.

j4james replied:
OK. If I make the update discussed above, I'll do this at the same time.

lhecker replied:
If you wonder about this: It's just because the former doesn't fit into a single register (16 bytes large) and the latter two do (8 bytes). Technically I wouldn't do such a teeny tiny micro-optimization, but MSVC has sort of a bug, where 16-byte large types are spilled on the stack for parameter passing using SSE instructions (`movdqu`/`movdqa`). Each function call is then not `2.5ns` but rather about `2.5ns + 10ns * number_of_16_byte_args`. In case of the VT parser this infamously led to a ~40% _overall_ speed up of Windows Terminal on my AMD Zen 3 CPU just by not passing 16 byte `string_views` and instead pass a C-string pointer via a register argument. So... That's why I'm a bit careful about this one thing in particular now...

lhecker reviewed (approved):
Statistically this PR should contain bugs, but if it does, I couldn't find it.
I'm a bit worried about the new `AdaptDispatch` functions (`_ScrollRectVertically`, etc.), because while the new code looks correct, I failed to determine what the new code corresponds with in the old code.

j4james commented:
> I'm a bit worried about the new `AdaptDispatch` functions (`_ScrollRectVertically`, etc.), because while the new code looks correct, I failed to determine what the new code corresponds with in the old code.

They're basically simplified variants of the the [`ScrollRegion`](https://github.com/microsoft/terminal/blob/ce6ce50e7e9eeb2d7377527d9512da9d43674bc0/src/host/output.cpp#L338) function. That's a lot more complicated, because it handles movements in arbitrary directions, and performs various additional clipping calculations which we just don't need. But in the end it comes down to three steps:

1. Copying a rectangle from one point to another, which is performed in the `_CopyRectangle` function, and which is special cased for vertical scrolls [here](https://github.com/microsoft/terminal/blob/ce6ce50e7e9eeb2d7377527d9512da9d43674bc0/src/host/output.cpp#L85-L92), and the more complicated cases [here](https://github.com/microsoft/terminal/blob/ce6ce50e7e9eeb2d7377527d9512da9d43674bc0/src/host/output.cpp#L98-L112).
2. Notifying the renderer in the [`_ScrollScreen`](https://github.com/microsoft/terminal/blob/ce6ce50e7e9eeb2d7377527d9512da9d43674bc0/src/host/output.cpp#L272) function. That also performs accessibility notifications, but those are broken and unnecessary (see #12656).
3. Filling the blank area that was revealed by the scroll and resetting the line renditions, which occurs at the [bottom of the ScrollRegion function](https://github.com/microsoft/terminal/blob/ce6ce50e7e9eeb2d7377527d9512da9d43674bc0/src/host/output.cpp#L447-L461).

For our simpler vertical scrolling requirements those steps are implemented as follows:

1. A call to [`textBuffer.ScrollRows`](https://github.com/microsoft/terminal/blob/64b4a0f3bbf56bf1a7d6f689011669500cdadba7/src/terminal/adapter/adaptDispatch.cpp#L458).
2. A call to [`textBuffer.TriggerRedraw`](https://github.com/microsoft/terminal/blob/64b4a0f3bbf56bf1a7d6f689011669500cdadba7/src/terminal/adapter/adaptDispatch.cpp#L459).
3. Calls to [`_FillRect`](https://github.com/microsoft/terminal/blob/64b4a0f3bbf56bf1a7d6f689011669500cdadba7/src/terminal/adapter/adaptDispatch.cpp#L462-L468) and [`textBuffer.ResetLineRenditionRange`](https://github.com/microsoft/terminal/blob/64b4a0f3bbf56bf1a7d6f689011669500cdadba7/src/terminal/adapter/adaptDispatch.cpp#L471).

For horizontal scrolling it's slightly different:

1. We need to use more complicated [`WalkInBounds` approach](https://github.com/microsoft/terminal/blob/64b4a0f3bbf56bf1a7d6f689011669500cdadba7/src/terminal/adapter/adaptDispatch.cpp#L493-L503) for copying the rect.
2. `TriggerRedraw` isn't necessary, because the `textBuffer.Write` from step 1 takes care of that.
3. We still have a call to [`_FillRect`](https://github.com/microsoft/terminal/blob/64b4a0f3bbf56bf1a7d6f689011669500cdadba7/src/terminal/adapter/adaptDispatch.cpp#L506-L512), but we don't need to reset line renditions for horizontal scrolling.

lhecker commented:
Just in case, I've checked this PR for regressions in raw text/VT throughput and it's ok.
The binary is about 2kB smaller (without the use of PGO).

The tools I use btw are https://github.com/lhecker/benchcat and https://github.com/lhecker/rainbowbench.
(`benchcat` is a bit special, because I tried writing an application without CRT. It's only 8kB!)

zadjii-msft commented:
I'm cool merging this. I'm a big kid, I can deal with the conflicts in my own branches. Happy to get this one in and get you unblocked ☺️ 

That being said, literally every time I've tried pushing a merge fix to a branch from a fork, I'm botched it horribly.

j4james commented:
@DHowett I'm still at work at the moment, but I should be able to look at it later tonight. I was hoping to try and address the issues that Leonard raised at the same time, although they probably aren't essential if you want to get this over and done with.

DHowett commented on src/terminal/adapter/ut_adapter/Adapter.UnitTests.vcxproj:
```diff
@@ -28,6 +28,9 @@
     <ProjectReference Include="..\..\..\interactivity\base\lib\InteractivityBase.vcxproj">
       <Project>{06ec74cb-9a12-429c-b551-8562ec964846}</Project>
     </ProjectReference>
+    <ProjectReference Include="..\..\..\renderer\base\lib\base.vcxproj">
```
Taking a mental note to adjust the link line in the `sources` file to include ConRenderBase (thanks for calling this one to my attention!)

j4james replied:
I've done the merge and the other updates. Would you like me to do this too or should I leave it for you to fix? I'm not exactly sure what is required, but I thought maybe I just had to add back this line:

    $(WINCORE_OBJ_PATH)\console\open\src\renderer\base\lib\$(O)\ConRenderBase.lib \

Or is there more to it than that?

DHowett replied:
There's a nonzero chance there's more to it than that (Base might depend on Interactivity, which would bring back in the whole kit&kaboodle). I'm happy to handle this one, but thank you!

DHowett replied:
(The best way to know is to try and build it, and the best way to do that is _probably having an actual Windows enlistment_, and.. well, you know.)

DHowett reviewed (commented):
I'm at 26/28, and I want to go get lunch. I'll swing back by for the other 2. I'm not surprised that I only have one comment so far -- this is top-notch work, as always. :)

DHowett commented:
> I was hoping to try and address the issues that Leonard raised at the same time

I'm happy to wait for this! Thanks!

DHowett commented on src/terminal/adapter/adaptDispatch.cpp:
```diff
@@ -353,12 +389,14 @@ bool AdaptDispatch::CursorRestoreState()
-    if (savedCursorState.IsOriginModeRelative && _scrollMargins.Bottom != 0)
+    // If the origin mode is relative, we need to make sure the restored
+    // position is clamped within the margins.
+    if (savedCursorState.IsOriginModeRelative)
```
I'm curious why we don't need to check any longer that there _are_ margins -- what overarching change made this safe to do both with and without margins?

j4james replied:
Now we have the `_GetVerticalMargins` helper function which always returns a valid set of margins - if margins haven't been set, it'll return the margins of the viewport itself. In this particular case it means we'd be clamping the row to the viewport, which is essentially a noop. But in general it makes things cleaner not having to care whether the margins are set or not.

DHowett replied:
Ah, right -- you called this out in the PR body. Thanks!

DHowett commented on src/terminal/adapter/adaptDispatch.cpp:
```diff
@@ -394,60 +432,100 @@ bool AdaptDispatch::CursorRestoreState()
-    // This uses a private API instead of the public one, because the public API
-    //      will set the cursor shape back to legacy.
-    _pConApi->SetCursorVisibility(fIsVisible);
+    _pConApi->GetTextBuffer().GetCursor().SetIsVisible(fIsVisible);
```
Might be good to ruminate on what happens if we ever make the Cursor a "global good" (since we learned that an alt buffer and main buffer share cursor properties, for example -- does it make sense to have one per text buffer then?)

j4james replied:
The problem is that the cursor coordinates are absolute, so when the buffers aren't the same size, the coordinates won't be the same. And when it comes to the console buffers, I think their cursor positions might be completely independent - I can't remember now.

Cursor properties (like visibility, shape, blinking status) could possibly be made global though. I've made a few attempts at that in the past, but always got stuck on something. I haven't given up on the idea, but it wasn't as straightforward as I expected it to be.

DHowett replied:
That totally makes sense. I legitimately forgot that the cursor's _position_ wasn't shared.

DHowett reviewed (commented):
I am so on board with this. I love the reduced indirection and the improved _obviousness_ of what adaptDispatch is doing. Thanks so much, James.

I'm only marking a comment here so that we don't automerge, but... this is an approval through and through

DHowett commented:
<img width="346" alt="awaiting approval from @j4james" src="https://user-images.githubusercontent.com/189190/163287201-5703c095-75f5-41c8-87a5-222c876010af.png">


Bot is going crazy. I'll fix that.

@msftbot forget everything i just told you

DHowett commented:
@msftbot nevermind
